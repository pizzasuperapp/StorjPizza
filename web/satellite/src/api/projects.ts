// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ErrorUnauthorized } from '@/api/errors/ErrorUnauthorized';
import {
    DataStamp,
    Project,
    ProjectFields,
    ProjectLimits,
    ProjectsApi,
    ProjectsCursor,
    ProjectsPage,
    ProjectsStorageBandwidthDaily,
} from '@/types/projects';
import { HttpClient } from '@/utils/httpClient';
import { Time } from "@/utils/time";

export class ProjectsApiGql extends BaseGql implements ProjectsApi {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/projects';

    /**
     * Creates project.
     *
     * @param projectFields - contains project information
     * @throws Error
     */
    public async create(projectFields: ProjectFields): Promise<Project> {
        const query =
            `mutation($name: String!, $description: String!) {
                createProject(
                    input: {
                        name: $name,
                        description: $description,
                    }
                ) {id}
            }`;

        const variables = {
            name: projectFields.name,
            description: projectFields.description,
        };

        const response = await this.mutate(query, variables);

        return new Project(response.data.createProject.id, variables.name, variables.description, '', projectFields.ownerId);
    }

    /**
     * Fetch projects.
     *
     * @returns Project[]
     * @throws Error
     */
    public async get(): Promise<Project[]> {
        const query = `query {
            myProjects{
                name
                id
                description
                createdAt
                ownerId
            }
        }`;

        const response = await this.query(query);

        return response.data.myProjects.map((project: Project) => {
            return new Project(
                project.id,
                project.name,
                project.description,
                project.createdAt,
                project.ownerId,
            );
        });
    }

    /**
     * Update project name and description.
     *
     * @param projectId - project ID
     * @param projectFields - project fields
     * @param projectLimits - project limits
     * @returns Project[]
     * @throws Error
     */
    public async update(projectId: string, projectFields: ProjectFields, projectLimits: ProjectLimits): Promise<void> {
        const query =
            `mutation($projectId: String!, $name: String!, $description: String!, $storageLimit: String!, $bandwidthLimit: String!) {
                updateProject(
                    id: $projectId,
                    projectFields: {
                        name: $name,
                        description: $description,
                    },
                    projectLimits: {
                        storageLimit: $storageLimit,
                        bandwidthLimit: $bandwidthLimit,
                    }
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
            name: projectFields.name,
            description: projectFields.description,
            storageLimit: projectLimits.storageLimit.toString(),
            bandwidthLimit: projectLimits.bandwidthLimit.toString(),
        };

        await this.mutate(query, variables);
    }

    /**
     * Delete project.
     *
     * @param projectId - project ID
     * @throws Error
     */
    public async delete(projectId: string): Promise<void> {
        const query =
            `mutation($projectId: String!) {
                deleteProject(
                    id: $projectId
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
        };

        await this.mutate(query, variables);
    }

    /**
     * Get project limits.
     *
     * @param projectId- project ID
     * throws Error
     */
    public async getLimits(projectId: string): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/${projectId}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
                limits.objectCount,
                limits.segmentCount,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get usage limits');
    }

    /**
     * Get total limits for all the projects that user owns.
     *
     * throws Error
     */
    public async getTotalLimits(): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get total usage limits');
    }

    /**
     * Get project daily usage for specific date range.
     *
     * @param projectId- project ID
     * @param start- since date
     * @param end- before date
     * throws Error
     */
    public async getDailyUsage(projectId: string, start: Date, end: Date): Promise<ProjectsStorageBandwidthDaily> {
        const since = Time.toUnixTimestamp(start).toString();
        const before = Time.toUnixTimestamp(end).toString();
        const path = `${this.ROOT_PATH}/${projectId}/daily-usage?from=${since}&to=${before}`;
        const response = await this.http.get(path);

        if (response.ok) {
            const usage = await response.json();

            return new ProjectsStorageBandwidthDaily(
                usage.storageUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
                usage.allocatedBandwidthUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
                usage.settledBandwidthUsage.map(el => {
                    const date = new Date(el.date)
                    date.setHours(0, 0, 0, 0)
                    return new DataStamp(el.value, date)
                }),
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get project daily usage');
    }

    /**
     * Fetch owned projects.
     *
     * @returns ProjectsPage
     * @throws Error
     */
    public async getOwnedProjects(cursor: ProjectsCursor): Promise<ProjectsPage> {
        const query =
            `query($limit: Int!, $page: Int!) {
                ownedProjects( cursor: { limit: $limit, page: $page } ) {
                    projects {
                        id,
                        name,
                        ownerId,
                        description,
                        createdAt,
                        memberCount
                    },
                    limit,
                    offset,
                    pageCount,
                    currentPage,
                    totalCount
                 }
             }`;

        const variables = {
            limit: cursor.limit,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.getProjectsPage(response.data.ownedProjects);
    }

    /**
     * Method for mapping projects page from json to ProjectsPage type.
     *
     * @param page anonymous object from json
     */
    private getProjectsPage(page: any): ProjectsPage { // eslint-disable-line @typescript-eslint/no-explicit-any
        if (!page) {
            return new ProjectsPage();
        }

        const projects: Project[] = page.projects.map(key =>
            new Project(
                key.id,
                key.name,
                key.description,
                key.createdAt,
                key.ownerId,
                false,
                key.memberCount));

        return new ProjectsPage(projects, page.limit, page.offset, page.pageCount, page.currentPage, page.totalCount);
    }

}
