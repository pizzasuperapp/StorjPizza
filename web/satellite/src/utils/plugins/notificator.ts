// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { NOTIFICATION_ACTIONS } from '@/utils/constants/actionNames';

interface Dispatcher {
    dispatch(key:string, message:string)
}

/**
 * Exposes UI notifications functionality.
 */
export class Notificator {
    public constructor(private store: Dispatcher) {}

    public async success(message: string): Promise<void> {
        await this.store.dispatch(NOTIFICATION_ACTIONS.SUCCESS, message);
    }

    public async error(message: string): Promise<void> {
        await this.store.dispatch(NOTIFICATION_ACTIONS.ERROR, message);
    }

    public async notify(message: string): Promise<void> {
        await this.store.dispatch(NOTIFICATION_ACTIONS.NOTIFY, message);
    }

    public async warning(message: string): Promise<void> {
        await this.store.dispatch(NOTIFICATION_ACTIONS.WARNING, message);
    }
}

/**
 * Registers plugin in Vue instance.
 */
export class NotificatorPlugin {
    constructor(private store: Dispatcher) {}
    public install(localVue: { prototype: { $notify: Notificator } }): void {
        localVue.prototype.$notify = new Notificator(this.store);
    }
}
