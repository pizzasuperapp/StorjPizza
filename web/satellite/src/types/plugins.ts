// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { Notificator } from '@/utils/plugins/notificator';

declare module 'vue/types/vue' {
    interface Vue {
        $notify: Notificator;
    }
}
