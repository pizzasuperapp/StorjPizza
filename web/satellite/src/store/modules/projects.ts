// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    DataStamp,
    Project,
    ProjectFields,
    ProjectLimits,
    ProjectsApi,
    ProjectsCursor,
    ProjectsPage,
    ProjectsStorageBandwidthDaily,
    ProjectUsageDateRange,
} from '@/types/projects';
import {StoreModule} from "@/types/store";

export const PROJECTS_ACTIONS = {
    FETCH: 'fetchProjects',
    FETCH_OWNED: 'fetchOwnedProjects',
    FETCH_DAILY_DATA: 'fetchDailyData',
    CREATE: 'createProject',
    CREATE_DEFAULT_PROJECT: 'createDefaultProject',
    SELECT: 'selectProject',
    UPDATE_NAME: 'updateProjectName',
    UPDATE_DESCRIPTION: 'updateProjectDescription',
    UPDATE_STORAGE_LIMIT: 'updateProjectStorageLimit',
    UPDATE_BANDWIDTH_LIMIT: 'updateProjectBandwidthLimit',
    DELETE: 'deleteProject',
    CLEAR: 'clearProjects',
    GET_LIMITS: 'getProjectLimits',
    GET_TOTAL_LIMITS: 'getTotalLimits',
};

export const PROJECTS_MUTATIONS = {
    ADD: 'CREATE_PROJECT',
    REMOVE: 'DELETE_PROJECT',
    UPDATE_PROJECT_NAME: 'UPDATE_PROJECT_NAME',
    UPDATE_PROJECT_DESCRIPTION: 'UPDATE_PROJECT_DESCRIPTION',
    UPDATE_PROJECT_STORAGE_LIMIT: 'UPDATE_STORAGE_LIMIT',
    UPDATE_PROJECT_BANDWIDTH_LIMIT: 'UPDATE_BANDWIDTH_LIMIT',
    SET_PROJECTS: 'SET_PROJECTS',
    SELECT_PROJECT: 'SELECT_PROJECT',
    CLEAR_PROJECTS: 'CLEAR_PROJECTS',
    SET_LIMITS: 'SET_PROJECT_LIMITS',
    SET_TOTAL_LIMITS: 'SET_TOTAL_LIMITS',
    SET_PAGE_NUMBER: 'SET_PAGE_NUMBER',
    SET_PAGE: 'SET_PAGE',
    SET_DAILY_DATA: 'SET_DAILY_DATA',
    SET_CHARTS_DATE_RANGE: 'SET_CHARTS_DATE_RANGE',
};

const defaultSelectedProject = new Project('', '', '', '', '', true, 0);

export class ProjectsState {
    public projects: Project[] = [];
    public selectedProject: Project = defaultSelectedProject;
    public currentLimits: ProjectLimits = new ProjectLimits();
    public totalLimits: ProjectLimits = new ProjectLimits();
    public cursor: ProjectsCursor = new ProjectsCursor();
    public page: ProjectsPage = new ProjectsPage();
    public allocatedBandwidthChartData: DataStamp[] = [];
    public settledBandwidthChartData: DataStamp[] = [];
    public storageChartData: DataStamp[] = [];
    public chartDataSince: Date = new Date();
    public chartDataBefore: Date = new Date();
}

interface ProjectsContext {
    state: ProjectsState
    commit: (string, ...unknown) => void
    rootGetters: {
        user: {
            id: string
        }
    }
    dispatch: (string, ...unknown) => Promise<any> // eslint-disable-line @typescript-eslint/no-explicit-any
}

const {
    FETCH,
    FETCH_DAILY_DATA,
    CREATE,
    CREATE_DEFAULT_PROJECT,
    SELECT,
    UPDATE_NAME,
    UPDATE_DESCRIPTION,
    UPDATE_STORAGE_LIMIT,
    UPDATE_BANDWIDTH_LIMIT,
    DELETE,
    CLEAR,
    GET_LIMITS,
    GET_TOTAL_LIMITS,
    FETCH_OWNED,
} = PROJECTS_ACTIONS;

const {
    ADD,
    REMOVE,
    UPDATE_PROJECT_NAME,
    UPDATE_PROJECT_DESCRIPTION,
    UPDATE_PROJECT_STORAGE_LIMIT,
    UPDATE_PROJECT_BANDWIDTH_LIMIT,
    SET_PROJECTS,
    SELECT_PROJECT,
    CLEAR_PROJECTS,
    SET_LIMITS,
    SET_TOTAL_LIMITS,
    SET_PAGE_NUMBER,
    SET_PAGE,
    SET_DAILY_DATA,
    SET_CHARTS_DATE_RANGE,
} = PROJECTS_MUTATIONS;
const projectsPageLimit = 7;

export function makeProjectsModule(api: ProjectsApi): StoreModule<ProjectsState, ProjectsContext> {
    return {
        state: new ProjectsState(),
        mutations: {
            [ADD](state: ProjectsState, createdProject: Project): void {
                state.projects.push(createdProject);
            },
            [SET_PROJECTS](state: ProjectsState, projects: Project[]): void {
                state.projects = projects;

                if (!state.selectedProject.id) {
                    return;
                }

                const projectsCount = state.projects.length;

                for (let i = 0; i < projectsCount; i++) {
                    const project = state.projects[i];

                    if (project.id !== state.selectedProject.id) {
                        continue;
                    }

                    state.selectedProject = project;

                    return;
                }

                state.selectedProject = defaultSelectedProject;
            },
            [SELECT_PROJECT](state: ProjectsState, projectID: string): void {
                const selected = state.projects.find((project: Project) => project.id === projectID);

                if (!selected) {
                    return;
                }

                state.selectedProject = selected;
            },
            [UPDATE_PROJECT_NAME](state: ProjectsState, fieldsToUpdate: ProjectFields): void {
                state.selectedProject.name = fieldsToUpdate.name;
            },
            [UPDATE_PROJECT_DESCRIPTION](state: ProjectsState, fieldsToUpdate: ProjectFields): void {
                state.selectedProject.description = fieldsToUpdate.description;
            },
            [UPDATE_PROJECT_STORAGE_LIMIT](state: ProjectsState, limitsToUpdate: ProjectLimits): void {
                state.currentLimits.storageLimit = limitsToUpdate.storageLimit;
            },
            [UPDATE_PROJECT_BANDWIDTH_LIMIT](state: ProjectsState, limitsToUpdate: ProjectLimits): void {
                state.currentLimits.bandwidthLimit = limitsToUpdate.bandwidthLimit;
            },
            [REMOVE](state: ProjectsState, projectID: string): void {
                state.projects = state.projects.filter(project => project.id !== projectID);

                if (state.selectedProject.id === projectID) {
                    state.selectedProject = new Project();
                }
            },
            [SET_LIMITS](state: ProjectsState, limits: ProjectLimits): void {
                state.currentLimits = limits;
            },
            [SET_TOTAL_LIMITS](state: ProjectsState, limits: ProjectLimits): void {
                state.totalLimits = limits;
            },
            [CLEAR_PROJECTS](state: ProjectsState): void {
                state.projects = [];
                state.selectedProject = defaultSelectedProject;
                state.currentLimits = new ProjectLimits();
                state.totalLimits = new ProjectLimits();
                state.storageChartData = [];
                state.allocatedBandwidthChartData = [];
                state.settledBandwidthChartData = [];
                state.chartDataSince = new Date();
                state.chartDataBefore = new Date();
            },
            [SET_PAGE_NUMBER](state: ProjectsState, pageNumber: number) {
                state.cursor.page = pageNumber;
                state.cursor.limit = projectsPageLimit;
            },
            [SET_PAGE](state: ProjectsState, page: ProjectsPage) {
                state.page = page;
            },
            [SET_DAILY_DATA](state: ProjectsState, payload: ProjectsStorageBandwidthDaily) {
                state.allocatedBandwidthChartData = payload.allocatedBandwidth;
                state.settledBandwidthChartData = payload.settledBandwidth;
                state.storageChartData = payload.storage;
            },
            [SET_CHARTS_DATE_RANGE](state: ProjectsState, payload: ProjectUsageDateRange) {
                state.chartDataSince = payload.since;
                state.chartDataBefore = payload.before;
            },
        },
        actions: {
            [FETCH]: async function ({commit}: ProjectsContext): Promise<Project[]> {
                const projects = await api.get();

                commit(SET_PROJECTS, projects);

                return projects;
            },
            [FETCH_OWNED]: async function ({commit, state}: ProjectsContext, pageNumber: number): Promise<ProjectsPage> {
                commit(SET_PAGE_NUMBER, pageNumber);

                const projectsPage: ProjectsPage = await api.getOwnedProjects(state.cursor);
                commit(SET_PAGE, projectsPage);

                return projectsPage;
            },
            [FETCH_DAILY_DATA]: async function ({commit, state}: ProjectsContext, payload: ProjectUsageDateRange): Promise<void> {
                const usage: ProjectsStorageBandwidthDaily = await api.getDailyUsage(state.selectedProject.id, payload.since, payload.before)

                commit(SET_CHARTS_DATE_RANGE, payload)
                commit(SET_DAILY_DATA, usage);
            },
            [CREATE]: async function ({commit}: ProjectsContext, createProjectFields: ProjectFields): Promise<Project> {
                const project = await api.create(createProjectFields);

                commit(ADD, project);

                return project;
            },
            [CREATE_DEFAULT_PROJECT]: async function ({rootGetters, dispatch}: ProjectsContext): Promise<void> {
                const UNTITLED_PROJECT_NAME = 'My First Project';
                const UNTITLED_PROJECT_DESCRIPTION = '___';
                const project = new ProjectFields(
                    UNTITLED_PROJECT_NAME,
                    UNTITLED_PROJECT_DESCRIPTION,
                    rootGetters.user.id,
                );
                const createdProject = await dispatch(CREATE, project);

                await dispatch(SELECT, createdProject.id);
            },
            [SELECT]: function ({commit}: ProjectsContext, projectID: string): void {
                commit(SELECT_PROJECT, projectID);
            },
            [UPDATE_NAME]: async function ({commit, state}: ProjectsContext, fieldsToUpdate: ProjectFields): Promise<void> {
                const project = new ProjectFields(
                    fieldsToUpdate.name,
                    state.selectedProject.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    state.currentLimits.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    state.currentLimits.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_NAME, fieldsToUpdate);
            },
            [UPDATE_DESCRIPTION]: async function ({commit, state}: ProjectsContext, fieldsToUpdate: ProjectFields): Promise<void> {
                const project = new ProjectFields(
                    state.selectedProject.name,
                    fieldsToUpdate.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    state.currentLimits.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    state.currentLimits.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_DESCRIPTION, fieldsToUpdate);
            },
            [UPDATE_STORAGE_LIMIT]: async function ({commit, state}: ProjectsContext, limitsToUpdate: ProjectLimits): Promise<void> {
                const project = new ProjectFields(
                    state.selectedProject.name,
                    state.selectedProject.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    state.currentLimits.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    limitsToUpdate.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_STORAGE_LIMIT, limitsToUpdate);
            },
            [UPDATE_BANDWIDTH_LIMIT]: async function ({commit, state}: ProjectsContext, limitsToUpdate: ProjectLimits): Promise<void> {
                const project = new ProjectFields(
                    state.selectedProject.name,
                    state.selectedProject.description,
                    state.selectedProject.id,
                );
                const limit = new ProjectLimits(
                    limitsToUpdate.bandwidthLimit,
                    state.currentLimits.bandwidthUsed,
                    state.currentLimits.storageLimit,
                    state.currentLimits.storageUsed,
                );
                await api.update(state.selectedProject.id, project, limit);

                commit(UPDATE_PROJECT_BANDWIDTH_LIMIT, limitsToUpdate);
            },
            [DELETE]: async function ({commit}: ProjectsContext, projectID: string): Promise<void> {
                await api.delete(projectID);

                commit(REMOVE, projectID);
            },
            [GET_LIMITS]: async function ({commit}: ProjectsContext, projectID: string): Promise<ProjectLimits> {
                const limits = await api.getLimits(projectID);

                commit(SET_LIMITS, limits);

                return limits;
            },
            [GET_TOTAL_LIMITS]: async function ({commit}: ProjectsContext): Promise<ProjectLimits> {
                const limits = await api.getTotalLimits();

                commit(SET_TOTAL_LIMITS, limits);

                return limits;
            },
            [CLEAR]: function({commit}: ProjectsContext): void {
                commit(CLEAR_PROJECTS);
            },
        },
        getters: {
            projects: (state: ProjectsState): Project[] => {
                return state.projects.map((project: Project) => {
                    if (project.id === state.selectedProject.id) {
                        project.isSelected = true;
                    }

                    return project;
                });
            },
            projectsWithoutSelected: (state: ProjectsState): Project[] => {
                return state.projects.filter((project: Project) => {
                    return project.id !== state.selectedProject.id;
                });
            },
            selectedProject: (state: ProjectsState): Project => state.selectedProject,
            projectsCount: (state: ProjectsState, rootGetters: ProjectsContext["rootGetters"]): number => {
                let projectsCount = 0;

                state.projects.forEach((project: Project) => {
                    if (project.ownerId === rootGetters.user.id) {
                        projectsCount++;
                    }
                });

                return projectsCount;
            },
        },
    };
}
