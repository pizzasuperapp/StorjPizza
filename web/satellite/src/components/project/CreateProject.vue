// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="create-project">
        <div class="create-project__container">
            <div class="create-project__container__image-container">
                <img
                    class="create-project__container__image-container__img"
                    src="@/../static/images/project/createProject.png"
                    alt="create project"
                >
            </div>
            <h2 class="create-project__container__title" aria-roledescription="title">Create a Project</h2>
            <VInput
                label="Project Name"
                additional-label="Up To 20 Characters"
                placeholder="Enter Project Name"
                class="full-input"
                is-limit-shown="true"
                :current-limit="projectName.length"
                :max-symbols="20"
                :error="nameError"
                @setData="setProjectName"
            />
            <VInput
                label="Description"
                placeholder="Enter Project Description"
                additional-label="Optional"
                class="full-input"
                is-multiline="true"
                height="100px"
                is-limit-shown="true"
                :current-limit="description.length"
                :max-symbols="100"
                @setData="setProjectDescription"
            />
            <div class="create-project__container__button-container">
                <VButton
                    class="create-project__container__button-container__cancel"
                    label="Cancel"
                    width="210px"
                    height="48px"
                    :on-press="onCancelClick"
                    is-transparent="true"
                />
                <VButton
                    label="Create Project +"
                    width="210px"
                    height="48px"
                    :on-press="onCreateProjectClick"
                    :is-disabled="!projectName"
                />
            </div>
            <div v-if="isLoading" class="create-project__container__blur">
                <VLoader class="create-project__container__blur__loader" width="50px" height="50px" />
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import VInput from '@/components/common/VInput.vue';
import VButton from '@/components/common/VButton.vue';
import VLoader from "@/components/common/VLoader.vue";

import { RouteConfig } from '@/router';
import { PROJECTS_ACTIONS } from '@/store/modules/projects';
import { ProjectFields } from '@/types/projects';
import { LocalData } from '@/utils/localData';

import { AnalyticsHttpApi } from '@/api/analytics';

// @vue/component
@Component({
    components: {
        VInput,
        VButton,
        VLoader,
    },
})
export default class NewProjectPopup extends Vue {
    private description = '';
    private createdProjectId = '';
    private isLoading = false;

    public projectName = '';
    public nameError = '';

    public readonly analytics: AnalyticsHttpApi = new AnalyticsHttpApi();

    /**
     * Sets project name from input value.
     */
    public setProjectName(value: string): void {
        this.projectName = value;
        this.nameError = '';
    }

    /**
     * Sets project description from input value.
     */
    public setProjectDescription(value: string): void {
        this.description = value;
    }

    /**
     * Redirects to previous route.
     */
    public onCancelClick(): void {
        const PREVIOUS_ROUTE_NUMBER = -1;

        this.$router.go(PREVIOUS_ROUTE_NUMBER);
    }

    /**
     * Creates project and refreshes store.
     */
    public async onCreateProjectClick(): Promise<void> {
        if (this.isLoading) {
            return;
        }

        this.isLoading = true;
        this.projectName = this.projectName.trim();

        const project = new ProjectFields(
            this.projectName,
            this.description,
            this.$store.getters.user.id,
        );

        try {
            project.checkName();
        } catch (error) {
            this.isLoading = false;
            this.nameError = error.message;

            return;
        }

        try {
            const createdProject = await this.$store.dispatch(PROJECTS_ACTIONS.CREATE, project);
            this.createdProjectId = createdProject.id;
        } catch (error) {
            this.isLoading = false;

            return;
        }

        this.selectCreatedProject();

        await this.$notify.success('Project created successfully!');

        this.isLoading = false;

        this.analytics.pageVisit(RouteConfig.ProjectDashboard.path);
        await this.$router.push(RouteConfig.ProjectDashboard.path);
    }

    /**
     * Selects just created project.
     */
    private selectCreatedProject(): void {
        this.$store.dispatch(PROJECTS_ACTIONS.SELECT, this.createdProjectId);
        LocalData.setSelectedProjectId(this.createdProjectId);
    }
}
</script>

<style scoped lang="scss">
    .create-project {
        width: 100%;
        height: calc(100% - 140px);
        padding: 70px 0;
        font-family: 'font_regular', sans-serif;

        &__container {
            margin: 0 auto;
            max-width: 440px;
            padding: 70px 50px 55px;
            background-color: #fff;
            border-radius: 8px;
            position: relative;

            &__image-container {
                width: 100%;
                display: flex;
                justify-content: center;
            }

            &__img {
                max-width: 190px;
                max-height: 130px;
            }

            &__title {
                font-size: 28px;
                line-height: 34px;
                color: #384b65;
                font-family: 'font_bold', sans-serif;
                text-align: center;
                margin: 15px 0 30px;
            }

            &__button-container {
                width: 100%;
                display: flex;
                align-items: center;
                justify-content: space-between;
                margin-top: 30px;

                &__cancel {
                    margin-right: 20px;
                }
            }

            &__blur {
                position: absolute;
                top: 0;
                left: 0;
                height: 100%;
                width: 100%;
                background-color: rgb(229 229 229 / 20%);
                border-radius: 8px;
                z-index: 100;

                &__loader {
                    width: 25px;
                    height: 25px;
                    position: absolute;
                    right: 40px;
                    top: 40px;
                }
            }
        }
    }

    .full-input {
        margin-top: 20px;
    }
</style>
