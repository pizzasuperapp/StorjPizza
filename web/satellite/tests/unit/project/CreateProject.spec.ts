// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import CreateProject from '@/components/project/CreateProject.vue';

import { makeProjectsModule } from '@/store/modules/projects';
import { NotificatorPlugin } from '@/utils/plugins/notificator';
import { createLocalVue, mount } from '@vue/test-utils';

import { ProjectsApiMock } from '../mock/api/projects';

const localVue = createLocalVue();

localVue.use(Vuex);

const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const store = new Vuex.Store({ modules: { projectsModule }});

localVue.use(new NotificatorPlugin(store));

describe('CreateProject.vue', (): void => {
    it('renders correctly', (): void => {
        const wrapper = mount<CreateProject>(CreateProject, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with project name', async (): Promise<void> => {
        const wrapper = mount<CreateProject>(CreateProject, {
            store,
            localVue,
        });

        await wrapper.vm.setProjectName('testName');

        expect(wrapper.findAll('.disabled').length).toBe(0);
    });
});
