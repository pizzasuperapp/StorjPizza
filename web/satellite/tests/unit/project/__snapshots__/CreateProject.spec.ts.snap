// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProject.vue renders correctly 1`] = `
<div class="create-project">
  <div class="create-project__container">
    <div class="create-project__container__image-container"><img src="@/../static/images/project/createProject.png" alt="create project" class="create-project__container__image-container__img"></div>
    <h2 aria-roledescription="title" class="create-project__container__title">Create a Project</h2>
    <div aria-roledescription="input-container" class="input-container full-input">
      <div class="label-container">
        <div class="label-container__main">
          <!---->
          <h3 class="label-container__main__label">Project Name</h3>
          <h3 class="label-container__main__label add-label">Up To 20 Characters</h3>
          <!---->
          <!---->
        </div>
        <h3 class="label-container__limit">0/20</h3>
      </div>
      <!---->
      <!----> <input id="Project Name" placeholder="Enter Project Name" maxlength="20" type="text" class="input" style="width: 100%; height: 48px;">
      <!---->
      <!---->
    </div>
    <div aria-roledescription="input-container" class="input-container full-input">
      <div class="label-container">
        <div class="label-container__main">
          <!---->
          <h3 class="label-container__main__label">Description</h3>
          <h3 class="label-container__main__label add-label">Optional</h3>
          <!---->
          <!---->
        </div>
        <h3 class="label-container__limit">0/100</h3>
      </div>
      <!----> <textarea id="Description" placeholder="Enter Project Description" rows="5" cols="40" maxlength="100" wrap="hard" class="textarea" style="width: 100%; height: 100px;"></textarea>
      <!---->
      <!---->
      <!---->
    </div>
    <div class="create-project__container__button-container">
      <div class="create-project__container__button-container__cancel container transparent" style="width: 210px; height: 48px; border-radius: 6px; font-size: 16px;">
        <!---->
        <!---->
        <!----> <span class="label">Cancel</span>
      </div>
      <div class="container disabled" style="width: 210px; height: 48px; border-radius: 6px; font-size: 16px;">
        <!---->
        <!---->
        <!----> <span class="label">Create Project +</span>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
